File: R/subsong.R
#' Get or set the current subsong in an `openmpt` module
#' 
#' Some `openmpt` modules may contain multiple subsongs. Use these
#' functions to get the current subsong index, or select a different one.
#' @param mod A tracker module object of class `openmpt`.
#' @param ... Ignored
#' @param value An `integer` index of the subsong to select.
#' @returns Returns the `integer` index of the currently selected subsong.
#' In case of the assign operator (`<-`) it returns a version of
#' `mod` with an update selection for the subsong
#' @examples
#' mod <- demo_mod()
#' subsong(mod)
#' ## a value of -1 plays all subsongs consecutively
#' subsong(mod) <- -1
#' @rdname subsong
#' @export
subsong <- function(mod, ...) {
  get_selected_subsong_(mod)
}

#' @rdname subsong
#' @export
`subsong<-` <- function(mod, ..., value) {
  select_subsong_(mod, as.integer(value))
}

File: src/subsong.cpp
#include <cpp11.hpp>
#include <libopenmpt/libopenmpt.hpp>
using namespace cpp11;
using namespace openmpt;

module * get_mod(SEXP mod); // specified in get_mod.cpp

[[cpp11::register]]
int get_selected_subsong_(SEXP mod) {
  module * my_mod = get_mod(mod);
  return my_mod->get_selected_subsong();
}

[[cpp11::register]]
SEXP select_subsong_(SEXP mod, int subsong) {
  module * my_mod = get_mod(mod);
  try {
    my_mod->select_subsong(subsong);
  } catch(...) {
    cpp11::stop("Failed to select subsong");
  }
  return mod;
}
